File: javascript/sdk-basics/ionic-ciphers/ionicCiphers.js
/*
 * (c) 2018-2020 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */
    
/*
 ** PRE-REQUISITES **
 * 1. A Machina tenant. You can obtain one at: https://ionic.com/start-for-free/.
 * 2. Your device, in this case the browser, needs to be enrolled. This is done 
 *    when you enrolled your device after signing up for a tenant. Enrollment 
 *    can also be accomplished by executing the Enroll Device script at:
 *    ../../enroll-device/index.html.
 */

import {getAgentConfig} from '../../jssdkConfig.js';

const main = async () => {
  
  // Get the tutorial application data. This assures all tutorials use the same
  // app ID, user ID and user authentication. It matches what was used for enrollment.
  const appData = getAgentConfig('Javascript Chunk Ciphers Tutorial');
  let response;

  // Initialize the Machina agent.
  try {
    response = await new window.IonicSdk.ISAgent(appData);
  } catch (errorResp) {
    console.error('Error initializing ionic agent:' + errorResp);
    return;
  }
  const agent = response.agent;

  /**********************************************************
   ** SENDER
   **********************************************************/
  const message = 'This is a secret message!';

  console.log('');
  console.log('Plain text:     ' + message);

  // encrypt data
  try {
    response = await agent.encryptStringChunkCipher({stringData: message});
  } catch (errorResp) {
    console.error('Error encrypting string:' + errorResp);
    return;
  }
  const cipherText = response.stringChunk;

  console.log('Sending Encrypted Text: ' + cipherText);

  /**********************************************************
   ** RECEIVER
   **********************************************************/
  console.log('-----------------------------------------------------------------------------------------------------');
  console.log('Receiving Encrypted Text: ' + cipherText);

  // decrypt data
  try {
    response = await agent.decryptStringChunkCipher({stringData: cipherText});
  } catch (errorResp) {
    console.error('Error decrypting:' + errorResp);
    return;
  }
  const decryptedText = response.stringChunk;
  console.log('Decrypted Text: ' + decryptedText);
};

main();

File: cpp/sdk-basics/ionic-ciphers/src/IonicCiphers.cpp
/*
 * (c) 2018-2020 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

#include "ISAgent.h"
#include "ISAgentSDKError.h"
#include <ISChunkCrypto.h>
#include <stdio.h>
#include <cstdlib>
#include <iostream>

#ifdef _WIN32
    #define HOMEVAR "USERPROFILE"
#else 
    #define HOMEVAR "HOME"
#endif

int main(int argc, char* argv[]) {

    int nErrorCode;

    // read persistor password from environment variable
    char* cpersistorPassword = std::getenv("IONIC_PERSISTOR_PASSWORD");
    if (cpersistorPassword == NULL) {
        std::cerr << "[!] Please provide the persistor password as env variable: IONIC_PERSISTOR_PASSWORD" << std::endl;
        exit(1);
    }
    std::string persistorPassword = std::string(cpersistorPassword);

    // initialize agent with password persistor
    std::string persistorPath = std::string(std::getenv(HOMEVAR)) + "/.ionicsecurity/profiles.pw";
    ISAgentDeviceProfilePersistorPassword persistor;
    persistor.setFilePath(persistorPath);
    persistor.setPassword(persistorPassword);
    ISAgent agent;
    nErrorCode = agent.initialize(persistor);
    if (nErrorCode != ISAGENT_OK) {
        std::cerr << "Failed to initialize agent from password persistor (" << persistorPath << ")" << std::endl;
        std::cerr << ISAgentSDKError::getErrorCodeString(nErrorCode) << std::endl;
        exit(1);
    }
    agent.setMetadata("ionic-application-name", "ionic-ciphers-tutorial");
    agent.setMetadata("ionic-application-version", "1.0.0");

    /******************************************************************************
     * SENDER
     ******************************************************************************/

    std::string message = "this is a secret message!";

    // initialize aes cipher object
    ISChunkCryptoCipherAuto  senderCipher(agent);

    // encrypt data
    std::string ciphertext;
    nErrorCode = senderCipher.encrypt(message, ciphertext);
    if (nErrorCode != ISCRYPTO_OK) {
        std::cerr << "Error: " << ISAgentSDKError::getErrorCodeString(nErrorCode) << std::endl;
        exit(1);
    }

    std::cout << "CIPHERTEXT    : " << ciphertext << std::endl;

    /******************************************************************************
     * RECEIVER
     ******************************************************************************/

    // initialize chunk cipher object
    ISChunkCryptoCipherAuto  receiverCipher(agent);

    // decrypt data
    std::string plaintext;
    receiverCipher.decrypt(ciphertext, plaintext);
    if (nErrorCode != ISCRYPTO_OK) {
        std::cerr << "Error: " << ISAgentSDKError::getErrorCodeString(nErrorCode) << std::endl;
        exit(1);
    }

    std::cout << "\nPLAINTEXT     : " << plaintext << std::endl;
}
